### PR TITLE
Add restrictions for wpcomvp user

### PR DESCRIPTION
**Pull request by sjinks**
## Description

This PR adds some restrictions for `wpcomvip` user:
* it does not show up in REST;
* it cannot be attributed any posts/pages/attachments;
* it is excluded from `wp_dropdown_users()`.

### Plugin Updated: VIP Helpers

Add restrictions to `wpcomvip` user to improve security.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [ ] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [x] I've created a changelog description that aligns with the provided examples.

## Steps to Test

The CI should pass.

### CODE REVIEW

File: vip-helpers/class-wpcomvip-restrictions.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		add_filter( 'wp_dropdown_users_args', [ $this, 'restrict_post_author_dropdown' ], PHP_INT_MAX );
+		add_filter( 'rest_user_query', [ $this, 'restrict_post_author_dropdown' ], PHP_INT_MAX );
+		add_filter( 'coauthors_edit_ignored_authors', [ $this, 'restrict_post_author_dropdown_coauthors' ], PHP_INT_MAX );
+	}
+
+	public function restrict_post_author( array $data ): array {
+		$wpcomvip = get_user_by( 'login', 'wpcomvip' );
+
+		if ( false !== $wpcomvip && isset( $data['post_author'] ) && $data['post_author'] == $wpcomvip->ID ) {
+			$data['post_author'] = get_current_user_id();
```
Comment (mjangda) [inline vip-helpers/class-wpcomvip-restrictions.php:28]: We probably want to check that current user is not `wpcomvip` either.

Comment (sjinks) [reply vip-helpers/class-wpcomvip-restrictions.php]: ✅

File: tests/vip-helpers/test-wpcomvip-restrictions.php [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+use WP_UnitTestCase;
+
+class Test_WPComVIP_Restrictions extends WP_UnitTestCase {
+	private static $user_id;
+	private static $post_id;
+
+	public static function wpSetUpBeforeClass( WP_UnitTest_Factory $factory ): void {
+		remove_all_filters( 'pre_get_lastpostmodified' );
+		remove_all_filters( 'transition_post_status' );
+		remove_all_filters( 'wpcom_vip_bump_lastpostmodified' );
```
Comment (sjinks) [inline tests/vip-helpers/test-wpcomvip-restrictions.php:20]: `lastpostmodified_Test` is sensitive to this :-(

**Comment by sjinks** [comment]
r2499-stacks

**Comment by jesse-greathouse** [comment]
Release v20220519.0

File: vip-helpers/class-wpcomvip-restrictions.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Automattic\VIP\Utils;
+
+class WPComVIP_Restrictions {
```
Comment (nickdaugherty) [inline vip-helpers/class-wpcomvip-restrictions.php:5]: Drive-by comment after-the-fact - since this class is already namespaced, we can drop the `WPComVIP_` prefix.

Comment (sjinks) [reply vip-helpers/class-wpcomvip-restrictions.php:5]: `wpcomvip` relates to the user :-)

File: vip-helpers/class-wpcomvip-restrictions.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	private function __construct() {
+		add_filter( 'wp_insert_post_data', [ $this, 'restrict_post_author' ], PHP_INT_MAX );
+		add_filter( 'wp_insert_attachment_data', [ $this, 'restrict_post_author' ], PHP_INT_MAX );
+		add_filter( 'wp_dropdown_users_args', [ $this, 'restrict_post_author_dropdown' ], PHP_INT_MAX );
```
Comment (WPprodigy) [inline vip-helpers/class-wpcomvip-restrictions.php:19]: The scope of `wp_dropdown_users_args` is more than just the author dropdown. I'm a little concerned it might mess with some jetpack logic wherein the `connection owner` is meant to be chosen by default for certain things. But our query filter will result in the connection owner never showing up.

We could target the post author dropdown specifically by checked the other args. Looks like `name = post_author_override` is unique.

Comment (sjinks) [reply vip-helpers/class-wpcomvip-restrictions.php:19]: Noted, thanks, I will take a look on Monday.
